File: problems/problems_3007/solution.py
import solution
from typing import *


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.findMaximumNumber(*test_input)

    def findMaximumNumber(self, k: int, x: int) -> int:
        num = pre_one = 0
        for i in range(((k + 1) << x).bit_length() - 1, -1, -1):
            # 当前i位为1的话，会增加:
            # 左边 2^i * pre_one 个价值
            # 右边 1到2^i-1里出现多少个x倍数位为1 的价值 (即 i//x * 2^(i-1)个1)
            # pre_one的价值会在遇到x倍数时增加
            cur = (pre_one << i) + (i // x << i >> 1)
            if cur <= k:
                k -= cur
                num |= 1 << i
                pre_one += (i + 1) % x == 0
        return num - 1

File: problems/problems_3007/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn find_maximum_number(mut k: i64, x: i32) -> i64 {
		let bit_length = |mut n: i64| {
			let mut res = 0;
			while n > 0 {
				n >>= 1;
				res += 1;
			}
			res
		};

		let mut num: i64 = 0;
		let mut pre1: i64 = 0;
		for i in (0..bit_length((k + 1) << x) - 1).rev() {
			let cur = (pre1 << i) + (((i / x) as i64) << i >> 1);
			if cur <= k {
				k -= cur;
				num |= 1 << i;
				if (i + 1) % x == 0 {
					pre1 += 1;
				}
			}
		}
		num - 1
    }
}

#[cfg(feature = "solution_3007")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let k: i64 = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	let x: i32 = serde_json::from_str(&input_values[1]).expect("Failed to parse input");
	json!(Solution::find_maximum_number(k, x))
}
